rpcs3/Emu/Cell/SPURecompiler.cpp

```diff
@@ -7202,6 +7202,32 @@ class spu_llvm_recompiler : public spu_recompiler_base, public cpu_translator
 		set_vr(op.rt, -(get_vr<f64[2]>(op.ra) * get_vr<f64[2]>(op.rb) + get_vr<f64[2]>(op.rt)));
 	}
 
+	// clamping helpers
+	value_t<f32[4]> clamp_positive_smax(value_t<f32[4]> v)
+	{
+		return eval(bitcast<f32[4]>(min(bitcast<s32[4]>(v),splat<s32[4]>(0x7f7fffff))));
+	}
+
+	value_t<f32[4]> clamp_negative_smax(value_t<f32[4]> v)
+	{
+		return eval(bitcast<f32[4]>(min(bitcast<u32[4]>(v),splat<u32[4]>(0xff7fffff))));
+	}
+
+	value_t<f32[4]> clamp_smax(value_t<f32[4]> v)
+	{
+		return eval(clamp_negative_smax(clamp_positive_smax(v)));
+	}
+
+	// FMA favouring zeros
+	value_t<f32[4]> xmuladd(value_t<f32[4]> a, value_t<f32[4]> b, value_t<f32[4]> c)
+	{
+		const auto ma = eval(sext<s32[4]>(fcmp_uno(a != fsplat<f32[4]>(0.))));
```

Comment:
Isn't it essentially a no-op here? Unordered comparison with zero. Maybe I'm confusing something, I don't get what cases it affects.

Reply:
This checks if the variable is not zero. We can distinguish 3 cases here:

- abs(a) < 0x7fffff => mask = 0 (because auf denormals deactivation)
- abs(a) between 0x80000 and 0x7f7fffffff => mask = ffffffff
- NaN => mask = fffffffff (in case someone feeds large constant values into the function)

Reply:
Yes, but why would it affect multiplication result?